File: services/backend/app/services/onboarding/chiffre_choc_selector.py
"""
Chiffre Choc Selector V2 — intention × lifecycle × confidence × available data.

Sprint S57 — ChiffreChoc V2.

Given a MinimalProfileResult and optional stress_type, selects the single
chiffre choc that will have the most impact on the user.

Selection hierarchy:
0. Critical archetype alerts (indep no LPP, expat low AVS)
1. Liquidity crisis (real data or severe)
2. Stress-aligned selection (if stress_type declared, data supports it)
3. Universal priorities (retirement gap, tax saving)
   — gated by lifecycle relevance and data confidence
4. Lifecycle-aware fallback (age-driven, always valid)

Confidence gating:
- If key data is estimated → confidence_mode = "pedagogical"
- If based on provided data or pure math → confidence_mode = "factual"

All text is in French, informal "tu", educational tone.

NEVER uses banned terms:
    "garanti", "certain", "assure", "sans risque",
    "optimal", "meilleur", "parfait", "conseiller",
    "tu devrais", "tu dois"

Sources:
    - LAVS art. 21-29 (rente AVS)
    - LPP art. 15-16 (bonifications vieillesse)
    - LIFD art. 38 (imposition du capital)
    - OPP3 art. 7 (plafond 3a)
"""

import math
from typing import Optional

from app.services.onboarding.onboarding_models import (
    ChiffreChoc,
    MinimalProfileResult,
)


# ═══════════════════════════════════════════════════════════════════════════════
# Compliance constants
# ═══════════════════════════════════════════════════════════════════════════════

_DISCLAIMER = (
    "Outil éducatif simplifié. Ne constitue pas un conseil financier (LSFin). "
    "Consulte un\u00b7e spécialiste pour une analyse personnalisée."
)

_SOURCES_RETIREMENT = [
    "LAVS art. 21-29 (rente AVS)",
    "LPP art. 15-16 (bonifications vieillesse)",
]

_SOURCES_TAX = [
    "OPP3 art. 7 (plafond 3a: 7'258 CHF)",
    "LIFD art. 33 (déduction 3a du revenu imposable)",
]

_SOURCES_LIQUIDITY = [
    "Recommandation générale: 3-6 mois de charges en réserve",
]

_SOURCES_COMPOUND = [
    "Calcul mathématique: intérêts composés à 3% nominal",
]


# ═══════════════════════════════════════════════════════════════════════════════
# Phase 0: Archetype alerts (aligned with Flutter _selectByArchetype)
# ═══════════════════════════════════════════════════════════════════════════════

def _select_by_archetype(profile: MinimalProfileResult) -> Optional[ChiffreChoc]:
    """Archetype-specific chiffre choc (highest priority when applicable).

    Aligned with Flutter ChiffreChocSelector._selectByArchetype.
    """
    # Independent without LPP: massive retirement gap alert
    if (profile.archetype == "independent_no_lpp"
        or (profile.estimated_replacement_ratio < 0.30
            and profile.projected_lpp_monthly <= 0
            and profile.gross_annual_salary > 0)):
        gap = max(0, profile.estimated_monthly_expenses - profile.estimated_monthly_retirement)
        lpp_estimated = "existing_lpp" in profile.estimated_fields
        return ChiffreChoc(
            category="retirement_gap",
            primary_number=round(gap, 0),
            display_text=(
                f"En tant qu'indépendant·e sans LPP, seule l'AVS te couvre. "
                f"Il te manquerait CHF {gap:,.0f} chaque mois à la retraite."
            ),
            explanation_text=(
                "Le 3e pilier (max CHF 36'288/an) et "
                "une LPP facultative peuvent combler cet écart."
            ),
            action_text="Découvre tes options de prévoyance \u2192",
            disclaimer=_DISCLAIMER,
            sources=list(_SOURCES_RETIREMENT),
            confidence_score=profile.confidence_score,
            confidence_mode="pedagogical" if lpp_estimated else "factual",
        )

    # Non-Swiss expat: AVS gap warning
    if (profile.archetype in ("expat_eu", "expat_non_eu")
            and profile.projected_avs_monthly < 1500):
        avs = profile.projected_avs_monthly
        is_eu = profile.archetype == "expat_eu"
        explanation = (
            "Tes années de cotisation en Europe comptent aussi grâce aux "
            "accords bilatéraux. Vérifie ta rente avec ton relevé CI."
            if is_eu else
            "Ta rente pourrait être réduite par des lacunes de cotisation. "
            "Demande ton relevé CI à ta caisse de compensation."
        )
        return ChiffreChoc(
            category="retirement_gap",
            primary_number=round(avs, 0),
            display_text=f"Ta rente AVS estimée: CHF {avs:,.0f}/mois.",
            explanation_text=explanation,
            action_text="Vérifie tes droits AVS \u2192",
            disclaimer=_DISCLAIMER,
            sources=list(_SOURCES_RETIREMENT),
            confidence_score=profile.confidence_score,
            confidence_mode="pedagogical",
        )

    return None


# ═══════════════════════════════════════════════════════════════════════════════
# Category builders
# ═══════════════════════════════════════════════════════════════════════════════

def _build_liquidity_choc(profile: MinimalProfileResult) -> ChiffreChoc:
    """Build chiffre choc for liquidity crisis (< 2 months runway)."""
    months = profile.months_liquidity
    monthly_expenses = profile.estimated_monthly_expenses

    display_text = (
        f"Ta réserve financière couvre environ {months:.1f} mois de charges. "
        f"Avec ~CHF {monthly_expenses:,.0f} de dépenses mensuelles estimées, "
        f"un imprévu pourrait vite devenir problématique."
    )
    explanation_text = (
        "Les spécialistes recommandent de conserver 3 à 6 mois de charges "
        "en épargne de précaution. En dessous de 2 mois, la marge de "
        "manœuvre est très réduite face à une perte d'emploi, une maladie "
        "ou une réparation urgente."
    )
    action_text = "Découvre comment constituer ta réserve de précaution pas à pas \u2192"

    savings_estimated = "current_savings" in profile.estimated_fields
    mode = "pedagogical" if savings_estimated else "factual"

    return ChiffreChoc(
        category="liquidity",
        primary_number=round(months, 1),
        display_text=display_text,
        explanation_text=explanation_text,
        action_text=action_text,
        disclaimer=_DISCLAIMER,
        sources=list(_SOURCES_LIQUIDITY),
        confidence_score=profile.confidence_score,
        confidence_mode=mode,
    )


def _build_retirement_gap_choc(profile: MinimalProfileResult) -> ChiffreChoc:
    """Build chiffre choc for retirement gap."""
    monthly_retirement = profile.estimated_monthly_retirement
    monthly_expenses = profile.estimated_monthly_expenses
    gap = max(0, monthly_expenses - monthly_retirement)

    display_text = (
        f"À la retraite, ton revenu mensuel estimé serait de "
        f"CHF {monthly_retirement:,.0f}. Aujourd'hui, tu dépenses "
        f"probablement ~CHF {monthly_expenses:,.0f} par mois."
    )

    if gap > 0:
        explanation_text = (
            f"Cela représente un écart d'environ CHF {gap:,.0f} par mois. "
            f"L'AVS et la LPP couvrent en moyenne 60% du dernier salaire. "
            f"Le 3e pilier et l'épargne libre permettent de combler ce gap."
        )
    else:
        explanation_text = (
            "Tes revenus projetés à la retraite semblent couvrir tes charges "
            "estimées. Toutefois, cette projection est basée sur des estimations "
            "simplifiées. Enrichis ton profil pour une analyse plus précise."
        )

    action_text = "Simule l'impact d'un 3e pilier sur ta situation \u2192"

    lpp_estimated = "existing_lpp" in profile.estimated_fields
    mode = "pedagogical" if lpp_estimated else "factual"

    return ChiffreChoc(
        category="retirement_gap",
        primary_number=round(gap, 0),
        display_text=display_text,
        explanation_text=explanation_text,
        action_text=action_text,
        disclaimer=_DISCLAIMER,
        sources=list(_SOURCES_RETIREMENT),
        confidence_score=profile.confidence_score,
        confidence_mode=mode,
    )


def _build_tax_saving_choc(profile: MinimalProfileResult) -> ChiffreChoc:
    """Build chiffre choc for tax saving opportunity via 3a."""
    tax_saving = profile.tax_saving_3a

    display_text = (
        f"En ouvrant un 3e pilier, tu pourrais économiser environ "
        f"CHF {tax_saving:,.0f} d'impôts chaque année."
    )
    explanation_text = (
        "Le versement au 3e pilier est déductible du revenu imposable. "
        "Avec un plafond de CHF 7'258 par an (salarié·e affilié·e LPP), "
        "l'économie fiscale dépend de ton taux marginal d'imposition."
    )
    action_text = "Explore les options 3a et leur impact fiscal \u2192"

    return ChiffreChoc(
        category="tax_saving",
        primary_number=round(tax_saving, 0),
        display_text=display_text,
        explanation_text=explanation_text,
        action_text=action_text,
        disclaimer=_DISCLAIMER,
        sources=list(_SOURCES_TAX),
        confidence_score=profile.confidence_score,
        confidence_mode="factual",  # Derived from salary + canton, both provided
    )


def _build_retirement_income_choc(profile: MinimalProfileResult) -> ChiffreChoc:
    """Build chiffre choc showing retirement income (positive framing)."""
    monthly_retirement = profile.estimated_monthly_retirement
    ratio_pct = round(profile.estimated_replacement_ratio * 100)

    display_text = (
        f"Avec l'AVS et la LPP, tu pourrais recevoir environ "
        f"CHF {monthly_retirement:,.0f} par mois à la retraite, "
        f"soit {ratio_pct}% de ton salaire actuel."
    )
    explanation_text = (
        "Ce taux de remplacement est dans la moyenne suisse. "
        "Le 3e pilier et l'épargne libre peuvent améliorer ta situation."
    )
    action_text = "Simule ta retraite en détail \u2192"

    lpp_estimated = "existing_lpp" in profile.estimated_fields
    mode = "pedagogical" if lpp_estimated else "factual"

    return ChiffreChoc(
        category="retirement_income",
        primary_number=round(monthly_retirement, 0),
        display_text=display_text,
        explanation_text=explanation_text,
        action_text=action_text,
        disclaimer=_DISCLAIMER,
        sources=list(_SOURCES_RETIREMENT),
        confidence_score=profile.confidence_score,
        confidence_mode=mode,
    )


def _build_compound_growth_choc(profile: MinimalProfileResult) -> ChiffreChoc:
    """Build compound growth choc for young users. Pure math, always factual."""
    age = profile.age  # A10 fix: no hasattr fallback — field is always present
    years = 65 - age
    monthly_contrib = 200.0
    annual_rate = 0.03
    monthly_rate = annual_rate / 12
    total_months = years * 12

    future_value = monthly_contrib * ((math.pow(1 + monthly_rate, total_months) - 1) / monthly_rate)

    reference_age = 35
    years_at_35 = 65 - reference_age
    months_at_35 = years_at_35 * 12
    future_at_35 = monthly_contrib * ((math.pow(1 + monthly_rate, months_at_35) - 1) / monthly_rate)

    advantage = future_value - future_at_35

    display_text = (
        f"200 CHF/mois dès maintenant = CHF {advantage:,.0f} de plus à 65 ans "
        f"qu'en commençant à 35."
    )
    explanation_text = (
        "Le temps est ton plus grand atout. Chaque année de cotisation supplémentaire "
        "profite des intérêts composés — ton argent travaille pour toi pendant "
        f"que tu vis ta vie. Sur {years} ans, même 200 CHF/mois font une différence majeure."
    )
    action_text = "Découvre combien ton 3a pourrait te rapporter \u2192"

    return ChiffreChoc(
        category="compound_growth",
        primary_number=round(advantage, 0),
        display_text=display_text,
        explanation_text=explanation_text,
        action_text=action_text,
        disclaimer=_DISCLAIMER,
        sources=list(_SOURCES_COMPOUND),
        confidence_score=profile.confidence_score,
        confidence_mode="factual",  # Pure math
    )


def _build_hourly_rate_choc(profile: MinimalProfileResult) -> ChiffreChoc:
    """Build hourly rate choc. Uses gross_annual_salary (provided), always factual.

    A2/A3 fix: uses profile.gross_annual_salary directly, NOT retirement proxy.
    """
    gross_annual = profile.gross_annual_salary
    working_hours_per_year = 2088.0
    net_annual = gross_annual * 0.75
    hourly_net = net_annual / working_hours_per_year if working_hours_per_year > 0 else 0

    rent_estimate = profile.estimated_monthly_expenses * 0.30
    rent_hours = round(rent_estimate / hourly_net) if hourly_net > 0 else 0

    display_text = (
        f"Après charges sociales et impôts, tu gagnes environ "
        f"CHF {hourly_net:.0f} de l'heure."
    )
    explanation_text = (
        f"Ton loyer te coûte ~{rent_hours} heures de travail par mois. "
        f"Savoir ce que tu gagnes vraiment par heure aide à évaluer "
        f"chaque dépense en temps de vie, pas seulement en francs."
    )
    action_text = "Explore ton budget en détail \u2192"

    return ChiffreChoc(
        category="hourly_rate",
        primary_number=round(hourly_net, 0),
        display_text=display_text,
        explanation_text=explanation_text,
        action_text=action_text,
        disclaimer=_DISCLAIMER,
        sources=[],
        confidence_score=profile.confidence_score,
        confidence_mode="factual",  # Pure math from provided salary
    )


# ═══════════════════════════════════════════════════════════════════════════════
# Stress-aligned selection
# ═══════════════════════════════════════════════════════════════════════════════

def _select_by_stress(
    stress_type: str, profile: MinimalProfileResult
) -> Optional[ChiffreChoc]:
    """Try to produce a chiffre choc aligned with user's declared intention.

    A3 fix: uses profile.gross_annual_salary for salary guard, not retirement proxy.
    """
    if stress_type == "stress_budget":
        # Budget: show hourly rate (pure math from salary — always factual)
        if profile.gross_annual_salary > 0:
            return _build_hourly_rate_choc(profile)
        return None

    if stress_type == "stress_impots":
        if profile.tax_saving_3a > 500:
            return _build_tax_saving_choc(profile)
        return None

    if stress_type == "stress_retraite":
        if profile.gross_annual_salary > 0:
            if profile.estimated_replacement_ratio < 0.55:
                return _build_retirement_gap_choc(profile)
            # Ratio OK → show retirement income (positive framing, not gap)
            return _build_retirement_income_choc(profile)
        return None

    # stress_patrimoine, stress_couple: no data at onboarding → fall through
    return None


# ═══════════════════════════════════════════════════════════════════════════════
# Lifecycle-aware fallback
# ═══════════════════════════════════════════════════════════════════════════════

def _select_by_lifecycle(profile: MinimalProfileResult) -> ChiffreChoc:
    """Lifecycle-aware fallback when no stress/priority matched."""
    age = profile.age

    if age < 28:
        return _build_compound_growth_choc(profile)

    if age < 38:
        if profile.existing_3a <= 0 and profile.tax_saving_3a > 1500:
            return _build_tax_saving_choc(profile)
        return _build_compound_growth_choc(profile)

    # 38+: retirement is relevant
    if profile.estimated_replacement_ratio < 0.55:
        return _build_retirement_gap_choc(profile)

    return _build_retirement_income_choc(profile)


# ═══════════════════════════════════════════════════════════════════════════════
# Main selector
# ═══════════════════════════════════════════════════════════════════════════════

def select_chiffre_choc(
    profile: MinimalProfileResult,
    stress_type: Optional[str] = None,
) -> ChiffreChoc:
    """Select the single most impactful chiffre choc for the user.

    V2: selection = intention × lifecycle × confidence × available data.

    Args:
        profile: The computed MinimalProfileResult.
        stress_type: Optional user intention ('stress_retraite', 'stress_budget', etc.)

    Returns:
        A single ChiffreChoc with category, display text, confidence_mode,
        and compliance fields.
    """
    # Phase 0: Archetype-specific alerts (A1 fix — was missing)
    archetype_choc = _select_by_archetype(profile)
    if archetype_choc is not None:
        return archetype_choc

    # Phase 1: Liquidity crisis — only if savings data is real or crisis is severe
    savings_estimated = "current_savings" in profile.estimated_fields
    if profile.months_liquidity < 2.0:
        if not savings_estimated or profile.months_liquidity < 1.0:
            return _build_liquidity_choc(profile)

    # Phase 2: Stress-aligned selection
    if stress_type and stress_type != "stress_general":
        stress_choc = _select_by_stress(stress_type, profile)
        if stress_choc is not None:
            return stress_choc

    # Phase 3: Universal priorities (gated by lifecycle)
    age = profile.age

    # Retirement gap: relevant from age 30+
    if age >= 30 and profile.estimated_replacement_ratio < 0.55:
        return _build_retirement_gap_choc(profile)

    # Tax saving 3a
    has_no_3a = profile.existing_3a <= 0.0
    if has_no_3a and profile.tax_saving_3a > 1500:
        return _build_tax_saving_choc(profile)

    # Phase 4: Lifecycle-aware fallback
    return _select_by_lifecycle(profile)

File: apps/mobile/lib/services/chiffre_choc_selector.dart
import 'dart:math';

import 'package:mint_mobile/models/minimal_profile_models.dart';
import 'package:mint_mobile/utils/chf_formatter.dart' as chf;

/// Selects the most impactful "chiffre choc" to show the user.
///
/// Sprint S57 — ChiffreChoc V2: intention × lifecycle × confidence × available data.
///
/// Selection hierarchy:
/// 1. Critical archetype alerts (indep no LPP, expat low AVS)
/// 2. Stress-aligned selection (if stressType declared, data supports it)
/// 3. Universal priorities (liquidity crisis, retirement gap, tax saving)
///    — gated by lifecycle relevance and data confidence
/// 4. Lifecycle-aware fallback (age-driven, always valid)
///
/// Confidence gating:
/// - If the chiffre choc's key data is estimated → [ChiffreChocConfidence.pedagogical]
/// - If based on provided data or pure math → [ChiffreChocConfidence.factual]
///
/// Legal basis: LAVS art. 21-40, LPP art. 7-16, OPP3 art. 7, LIFD art. 38.
class ChiffreChocSelector {
  ChiffreChocSelector._();

  /// Select the most impactful chiffre choc for the given profile.
  ///
  /// [stressType] — user's declared intention from StepStressSelector.
  /// When set and not 'stress_general', influences which type of chiffre
  /// choc is selected (if the data supports it).
  static ChiffreChoc select(
    MinimalProfileResult profile, {
    String? stressType,
  }) {
    // Phase 0: Critical archetype alerts — always override
    final archetypeChoc = _selectByArchetype(profile);
    if (archetypeChoc != null) return _withConfidence(archetypeChoc, profile);

    // Phase 1: Stress-aligned selection (NEW)
    if (stressType != null && stressType != 'stress_general') {
      final stressChoc = _selectByStress(stressType, profile);
      if (stressChoc != null) return _withConfidence(stressChoc, profile);
    }

    // Phase 2: Universal priorities — gated by lifecycle relevance
    // Liquidity alert: only when savings data is REAL (not estimated)
    // OR when the crisis is severe (< 1 month even with estimation)
    final savingsEstimated = profile.estimatedFields.contains('currentSavings');
    if (profile.liquidityMonths < 2 &&
        profile.currentSavings >= 0 &&
        (!savingsEstimated || profile.liquidityMonths < 1)) {
      return _withConfidence(_buildLiquidityChoc(profile), profile);
    }

    // Retirement gap: relevant from age 30+ (below that, it's too abstract)
    if (profile.age >= 30 &&
        profile.replacementRate < 0.55 &&
        profile.grossMonthlySalary > 0) {
      return _withConfidence(_buildRetirementGapChoc(profile), profile);
    }

    // Tax saving 3a: always relevant if applicable
    if (profile.existing3a <= 0 && profile.taxSaving3a > 1500) {
      return _withConfidence(_buildTaxSaving3aChoc(profile), profile);
    }

    // Phase 3: Lifecycle-aware fallback (NEW)
    return _withConfidence(_selectByLifecycle(profile), profile);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // Phase 1: Stress-aligned selection
  // ═══════════════════════════════════════════════════════════════════════════

  /// Try to produce a chiffre choc aligned with the user's declared intention.
  ///
  /// Returns null if the data doesn't support a meaningful choc for this stress.
  static ChiffreChoc? _selectByStress(
    String stressType,
    MinimalProfileResult profile,
  ) {
    switch (stressType) {
      case 'stress_budget':
        // Budget: show hourly rate (pure math from salary — always factual)
        if (profile.grossMonthlySalary > 0) {
          return _buildHourlyRateChoc(profile);
        }
        return null;

      case 'stress_impots':
        // Tax: show 3a tax saving if applicable
        if (profile.taxSaving3a > 500) {
          return _buildTaxSaving3aChoc(profile);
        }
        return null;

      case 'stress_retraite':
        // Retirement: show gap or income depending on data quality
        if (profile.grossMonthlySalary > 0) {
          if (profile.replacementRate < 0.55) {
            return _buildRetirementGapChoc(profile);
          }
          return _buildRetirementIncomeChoc(profile);
        }
        return null;

      case 'stress_patrimoine':
        // Patrimoine: only if we have real data — don't estimate fortune
        // Without real patrimoine data, fall through to lifecycle
        return null;

      case 'stress_couple':
        // Couple: no couple data at onboarding — fall through
        return null;

      default:
        return null;
    }
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // Phase 3: Lifecycle-aware fallback
  // ═══════════════════════════════════════════════════════════════════════════

  /// Lifecycle-aware fallback when no stress/universal priority matched.
  ///
  /// Uses pure math or well-grounded calculations per age group.
  /// Avoids showing retirement projections to users under 28.
  static ChiffreChoc _selectByLifecycle(MinimalProfileResult profile) {
    if (profile.age < 28) {
      // Young: compound growth advantage (pure math, no estimation)
      return _buildCompoundGrowthChoc(profile);
    }
    if (profile.age < 38) {
      // Construction: tax saving 3a is the most actionable lever
      if (profile.existing3a <= 0 && profile.taxSaving3a > 1500) {
        return _buildTaxSaving3aChoc(profile);
      }
      // Else: compound growth still meaningful
      return _buildCompoundGrowthChoc(profile);
    }
    // 38+: retirement income/gap is relevant
    if (profile.replacementRate < 0.55 && profile.grossMonthlySalary > 0) {
      return _buildRetirementGapChoc(profile);
    }
    return _buildRetirementIncomeChoc(profile);
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // Archetype alerts (unchanged priority 0)
  // ═══════════════════════════════════════════════════════════════════════════

  /// Archetype-specific chiffre choc (highest priority when applicable).
  static ChiffreChoc? _selectByArchetype(MinimalProfileResult profile) {
    // Independent without LPP: massive retirement gap alert
    if (profile.employmentStatus == 'independant' &&
        profile.lppMonthlyRente <= 0 &&
        profile.grossMonthlySalary > 0) {
      final gapFormatted = chf.formatChfWithPrefix(profile.retirementGapMonthly);
      final plafondStr = profile.plafond3a != null
          ? chf.formatChf(profile.plafond3a!)
          : '?';
      return ChiffreChoc(
        type: ChiffreChocType.retirementGap,
        value: '$gapFormatted/mois',
        rawValue: profile.retirementGapMonthly,
        title: 'Sans 2e pilier, ton gap de retraite',
        subtitle:
            'En tant qu\'independant\u00b7e sans LPP, seule l\'AVS te couvre. '
            'Il te manquerait $gapFormatted chaque mois a la retraite. '
            'Le 3e pilier (max CHF\u00A0$plafondStr/an) et '
            'une LPP facultative peuvent combler cet ecart.',
        iconName: 'warning_amber',
        colorKey: 'error',
      );
    }

    // Non-Swiss expat: AVS gap warning (only when nationality is known)
    if (profile.nationalityGroup != null &&
        profile.nationalityGroup != 'CH' &&
        profile.avsMonthlyRente < 1500) {
      final avsFormatted = chf.formatChfWithPrefix(profile.avsMonthlyRente);
      final subtitle = profile.nationalityGroup == 'EU'
          ? 'Tes annees de cotisation en Europe comptent aussi grace aux '
              'accords bilateraux. Avec $avsFormatted/mois d\'AVS estime, '
              'verifie ta rente avec ton releve de compte individuel (CI).'
          : 'Avec $avsFormatted/mois d\'AVS estime, ta rente pourrait etre '
              'reduite par des lacunes de cotisation. Demande ton releve CI '
              'a ta caisse de compensation.';
      return ChiffreChoc(
        type: ChiffreChocType.retirementGap,
        value: '$avsFormatted/mois',
        rawValue: profile.avsMonthlyRente,
        title: 'Ta rente AVS estimee',
        subtitle: subtitle,
        iconName: 'public',
        colorKey: 'warning',
      );
    }

    return null;
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // Chiffre choc builders
  // ═══════════════════════════════════════════════════════════════════════════

  static ChiffreChoc _buildLiquidityChoc(MinimalProfileResult profile) {
    return ChiffreChoc(
      type: ChiffreChocType.liquidityAlert,
      value: '${profile.liquidityMonths.toStringAsFixed(1)} mois',
      rawValue: profile.liquidityMonths,
      title: 'Ta reserve de liquidite',
      subtitle: profile.liquidityMonths < 1
          ? 'Moins d\'un mois de reserves. '
              'Les experts recommandent 3 a 6 mois de depenses en epargne de securite.'
          : '${profile.liquidityMonths.toStringAsFixed(1)} mois de reserves. '
              'Les experts recommandent 3 a 6 mois de depenses en epargne de securite.',
      iconName: 'warning_amber',
      colorKey: 'error',
    );
  }

  static ChiffreChoc _buildRetirementGapChoc(MinimalProfileResult profile) {
    final gapFormatted = chf.formatChfWithPrefix(profile.retirementGapMonthly);
    final pct = (profile.replacementRate * 100).round();
    final isIndep = profile.employmentStatus == 'independant';
    final plafond = profile.plafond3a;
    final subtitle = isIndep && plafond != null
        ? 'Sans 2e pilier obligatoire, ton ecart de retraite est plus important. '
            'Avec seulement $pct% de remplacement, il te manquerait $gapFormatted/mois. '
            'Le 3e pilier (max CHF\u00A0${chf.formatChf(plafond)}/an) est ton principal levier.'
        : '\u00c0 la retraite, tu pourrais recevoir environ $pct% de ton revenu actuel. '
            'Il te manquerait $gapFormatted chaque mois. '
            'Decouvre comment reduire cet ecart.';
    return ChiffreChoc(
      type: ChiffreChocType.retirementGap,
      value: '$gapFormatted/mois',
      rawValue: profile.retirementGapMonthly,
      title: 'Ton ecart de retraite',
      subtitle: subtitle,
      iconName: 'trending_down',
      colorKey: 'warning',
    );
  }

  static ChiffreChoc _buildTaxSaving3aChoc(MinimalProfileResult profile) {
    final savingFormatted = chf.formatChfWithPrefix(profile.taxSaving3a);
    final plafondText = profile.plafond3a != null
        ? chf.formatChf(profile.plafond3a!)
        : '?';
    return ChiffreChoc(
      type: ChiffreChocType.taxSaving3a,
      value: '$savingFormatted/an',
      rawValue: profile.taxSaving3a,
      title: 'Ton economie d\'impot potentielle',
      subtitle: 'En cotisant au 3e pilier (max CHF\u00A0$plafondText/an), '
          'tu pourrais economiser environ $savingFormatted d\'impots chaque annee. '
          'Et tu prepares ta retraite en meme temps.',
      iconName: 'savings',
      colorKey: 'success',
    );
  }

  static ChiffreChoc _buildRetirementIncomeChoc(MinimalProfileResult profile) {
    final retirementFormatted = chf.formatChfWithPrefix(profile.totalMonthlyRetirement);
    final pct = (profile.replacementRate * 100).round();
    return ChiffreChoc(
      type: ChiffreChocType.retirementIncome,
      value: '$retirementFormatted/mois',
      rawValue: profile.totalMonthlyRetirement,
      title: 'Ton revenu estime a la retraite',
      subtitle: 'Avec l\'AVS et la LPP, tu pourrais recevoir environ '
          '$retirementFormatted par mois, soit $pct% de ton salaire actuel.',
      iconName: 'account_balance',
      colorKey: 'info',
    );
  }

  /// Compound growth advantage for young users.
  ///
  /// Pure math: compares starting now vs starting at 35.
  /// Always [ChiffreChocConfidence.factual] — no estimation involved.
  static ChiffreChoc _buildCompoundGrowthChoc(MinimalProfileResult profile) {
    final years = 65 - profile.age;
    const monthlyContrib = 200.0;
    const annualRate = 0.03;
    const monthlyRate = annualRate / 12;
    final totalMonths = years * 12;

    // Future value of annuity: PMT × ((1 + r)^n - 1) / r
    final futureValue = monthlyContrib *
        ((pow(1 + monthlyRate, totalMonths) - 1) / monthlyRate);

    // Compare to starting at 35
    const referenceAge = 35;
    const yearsAt35 = 65 - referenceAge;
    const monthsAt35 = yearsAt35 * 12;
    final futureAt35 = monthlyContrib *
        ((pow(1 + monthlyRate, monthsAt35) - 1) / monthlyRate);

    final advantage = futureValue - futureAt35;
    final advantageFormatted = chf.formatChfWithPrefix(advantage);

    return ChiffreChoc(
      type: ChiffreChocType.compoundGrowth,
      value: advantageFormatted,
      rawValue: advantage,
      title: 'Ton avantage temps',
      subtitle: '200 CHF/mois des maintenant = $advantageFormatted de plus '
          'a 65\u00A0ans qu\'en commencant a 35. '
          'Le temps est ton plus grand atout.',
      iconName: 'trending_up',
      colorKey: 'success',
      confidenceMode: ChiffreChocConfidence.factual, // Pure math
    );
  }

  /// Net hourly rate breakdown.
  ///
  /// Pure math from provided salary — always factual.
  /// Shows what the user really earns per hour, making abstract salary concrete.
  static ChiffreChoc _buildHourlyRateChoc(MinimalProfileResult profile) {
    // Swiss standard: 42h/week × 52 weeks = 2'184h, minus 5 weeks vacation
    // → ~1'974 working hours/year. Simplified: 174h/month × 12 = 2'088h.
    const workingHoursPerYear = 2088.0;
    // Approximate net = 75% of gross (social charges + taxes)
    final netAnnual = profile.grossAnnualSalary * 0.75;
    final hourlyNet = netAnnual / workingHoursPerYear;
    final hourlyFormatted = 'CHF\u00A0${hourlyNet.round()}';

    // Housing cost in hours
    final monthlyExpenses = profile.estimatedMonthlyExpenses;
    // Estimate rent ~30% of expenses for budget-focused insight
    final rentEstimate = monthlyExpenses * 0.30;
    final rentHours = (rentEstimate / (hourlyNet)).round();

    return ChiffreChoc(
      type: ChiffreChocType.hourlyRate,
      value: '$hourlyFormatted/h',
      rawValue: hourlyNet,
      title: 'Ton salaire reel',
      subtitle: 'Apres charges sociales et impots, tu gagnes environ '
          '$hourlyFormatted de l\'heure. '
          'Ton loyer te coute ~$rentHours heures de travail par mois.',
      iconName: 'schedule',
      colorKey: 'info',
      confidenceMode: ChiffreChocConfidence.factual, // Derived from provided salary
    );
  }

  // ═══════════════════════════════════════════════════════════════════════════
  // Confidence gating
  // ═══════════════════════════════════════════════════════════════════════════

  /// Apply confidence mode based on whether key data for this choc is estimated.
  ///
  /// Rules:
  /// - [compoundGrowth] and [hourlyRate] are always factual (pure math)
  /// - [liquidityAlert] is pedagogical if currentSavings is estimated
  /// - [retirementGap] / [retirementIncome] are pedagogical if LPP is estimated
  /// - [taxSaving3a] is factual (derived from salary + canton, both provided)
  static ChiffreChoc _withConfidence(
    ChiffreChoc choc,
    MinimalProfileResult profile,
  ) {
    // Already set explicitly (e.g. compoundGrowth, hourlyRate)
    if (choc.confidenceMode != ChiffreChocConfidence.factual) return choc;

    final estimated = profile.estimatedFields;
    final ChiffreChocConfidence mode;

    switch (choc.type) {
      case ChiffreChocType.compoundGrowth:
      case ChiffreChocType.hourlyRate:
      case ChiffreChocType.taxSaving3a:
        mode = ChiffreChocConfidence.factual;
      case ChiffreChocType.liquidityAlert:
        mode = estimated.contains('currentSavings')
            ? ChiffreChocConfidence.pedagogical
            : ChiffreChocConfidence.factual;
      case ChiffreChocType.retirementGap:
      case ChiffreChocType.retirementIncome:
        mode = estimated.contains('existingLpp')
            ? ChiffreChocConfidence.pedagogical
            : ChiffreChocConfidence.factual;
    }

    if (mode == choc.confidenceMode) return choc;

    return ChiffreChoc(
      type: choc.type,
      value: choc.value,
      rawValue: choc.rawValue,
      title: choc.title,
      subtitle: choc.subtitle,
      iconName: choc.iconName,
      colorKey: choc.colorKey,
      confidenceMode: mode,
    );
  }

  // F3: _formatChf / _formatChfPlain removed — use centralized chf_formatter.dart
}
